services/debugger/cache/embeddedcache/embeddedCache.go

```diff
@@ -0,0 +1,192 @@
+	_, _ = fmt.Fprintf(os.Stderr, format, a...)
+}
+
+func (badgerLogger) Warningf(format string, a ...interface{}) {
+	_, _ = fmt.Fprintf(os.Stderr, format, a...)
+}
+
+// Update writes the entries into badger db with a TTL
+func (e *EmbeddedCache[E]) Update(key string, value E) {
+	e.Init()
+	txn := e.Db.NewTransaction(true)
+	res := e.Read(key)
+	if len(res) >= e.Limiter {
+		res = res[len(res)-e.Limiter+1:]
+	}
+	res = append(res, value)
+	data, err := json.Marshal(res)
+	if err != nil {
+		return
+	}
+	setBadgerEntry := func(ctx context.Context) error {
+		entry := badger.NewEntry([]byte(key), data).WithTTL(e.CleanupFreq)
+		if err := txn.SetEntry(entry); err == badger.ErrTxnTooBig {
+			err = txn.Commit()
+			if err != nil {
+				return err
+			}
+			txn = e.Db.NewTransaction(true)
+			err = txn.SetEntry(entry)
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+	_ = misc.RetryWith(context.TODO(), e.queryTimeout, e.retries, setBadgerEntry)
+	_ = txn.Commit()
+}
+
+// Read fetches all the entries for a given key from badgerDB
+func (e *EmbeddedCache[E]) Read(key string) []E {
+	e.Init()
+	read := func(ctx context.Context) ([]E, error) {
+		var values []E
+		err := e.Db.View(func(txn *badger.Txn) error {
+			it := txn.NewIterator(badger.DefaultIteratorOptions)
+			defer it.Close()
+			prefix := []byte(key)
+			for it.Seek(prefix); it.ValidForPrefix(prefix); it.Next() {
+				item := it.Item()
+				err := item.Value(func(v []byte) error {
+					err := json.Unmarshal(v, &values)
+					if err != nil {
+						return err
+					}
+					return nil
+				})
+				if err != nil {
+					return err
+				}
+			}
+			return nil
+		})
+		if err != nil {
+			return nil, err
+		}
+		return values, nil
+	}
+
+	res, _ := misc.QueryWithRetries(context.TODO(), e.queryTimeout, e.retries, read)
+
+	return res
+}
+
+func (e *EmbeddedCache[E]) Init() {
+	e.once.Do(func() {
+		e.loadCacheConfig()
+		badgerPathName := e.Origin + "/cache/badgerdbv3"
+		tmpDirPath, err := misc.CreateTMPDIR()
+		if err != nil {
+			e.Logger.Errorf("Unable to create tmp directory: %v", err)
+			return
+		}
+		storagePath := path.Join(tmpDirPath, badgerPathName)
+		e.path = storagePath
+		e.done = make(chan struct{})
+		e.closed = make(chan struct{})
+		opts := badger.
+			DefaultOptions(storagePath).
+			WithLogger(badgerLogger{e.Logger}).
+			WithCompression(options.None).
+			WithIndexCacheSize(16 << 20). // 16mb
+			WithNumGoroutines(1).
+			WithNumMemtables(0).
+			WithBlockCacheSize(0)
+
+		e.Db, err = badger.Open(opts)
+		if err != nil {
+			e.Logger.Errorf("Error while opening badgerDB: %v", err)
+			return
+		}
+		rruntime.Go(func() {
+			e.gcBadgerDB()
+		})
+	})
+}
+
+func (e *EmbeddedCache[E]) gcBadgerDB() {
+	ticker := time.NewTicker(e.Ticker)
+	defer ticker.Stop()
+	// One call would only result in removal of at max one log file.
+	// As an optimization, you could also immediately re-run it whenever it returns nil error
+	// (this is why `goto again` is used).
```

Comment:
This is true but it is also true that we should do this in periods of low activity as per Badger documentation. Usually these things are "Stop the world" events. This means that all activities on this Badger DB instance are stopped until the operation completes.
Since we're not running this during periods of low activity we can consider running it periodically without the `goto again`. @atzoum you played with badger more, any thoughs on this?

Reply:
This part of the code is copied from badgerDB documentation so I just took that as is

Reply:
I know but in the same docs where you got that they are suggesting to run it during periods of low activity.